src/bluetooth/qbluetoothtransferreply_bluez.cpp

```diff
@@ -91,12 +82,63 @@ QBluetoothTransferReplyBluez::QBluetoothTransferReplyBluez(QIODevice *input, con
 */
 QBluetoothTransferReplyBluez::~QBluetoothTransferReplyBluez()
 {
-    QDBusConnection::sessionBus().unregisterObject(m_agent_path);
+    delete objectTransfer;
+    delete objectPush;
     delete client;
 }
 
+void QBluetoothTransferReplyBluez::fail(QBluetoothTransferReply::TransferError err,
+	    const QString msg)
+{
+    m_finished = true;
+    m_running = false;
+    m_errorStr = msg;
+    m_error = err;
+    QMetaObject::invokeMethod(this, "finished",
+			      Qt::QueuedConnection,
+			      Q_ARG(QBluetoothTransferReply*, this));
+}
+
 bool QBluetoothTransferReplyBluez::start()
 {
+    QString dest = request().address().toString();
+    QVariantMap args;
+    args.insert(QString::fromLatin1("Target"), QString::fromLatin1("OPP"));
+
+    if (!objectPush) {
+        qCDebug(QT_BT_BLUEZ) << "Creating an outgoing session";
+	QDBusPendingReply<> sessionReply = client->CreateSession(dest, args);
+	QDBusPendingCallWatcher *watcher = new QDBusPendingCallWatcher(sessionReply, this);
+	QObject::connect(watcher, SIGNAL(finished(QDBusPendingCallWatcher*)),
+			 this, SLOT(sessionAcquired(QDBusPendingCallWatcher*)));
+    } else {
+        startPush();
+    }
+    return true;
+}
+
+void QBluetoothTransferReplyBluez::sessionAcquired(QDBusPendingCallWatcher *watcher)
+{
+    QDBusPendingReply<QDBusObjectPath> sessionReply = *watcher;
+    if (sessionReply.isError()) {
+        QString msg = sessionReply.error().message();
+        qCDebug(QT_BT_BLUEZ) << "Failed to create a session:" << msg;
+        fail(QBluetoothTransferReply::UnknownError, msg);
+	return;
+    }
+
+    qCDebug(QT_BT_BLUEZ) << "Creating a push object for path" << sessionReply.value().path();
+    delete objectPush;
+    objectPush = new OrgBluezObexObjectPushInterface
+        (QLatin1String("org.bluez.obex.client"),
```

Comment:
Use QStringLiteral(). It would also be worthwhile adding these as constants to the unnamed namespace.